File: ios/CareDoc/Sources/CareDoc/DesignSystem/Components.swift
import SwiftUI

// MARK: - 卡片

struct CardModifier: ViewModifier {
    var padding: CGFloat = 14
    var radius: CGFloat = CD.rCard
    func body(content: Content) -> some View {
        content
            .padding(padding)
            .background(CD.surface)
            .clipShape(RoundedRectangle(cornerRadius: radius, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: radius, style: .continuous)
                    .stroke(CD.cardBorder, lineWidth: 1)
            )
    }
}

extension View {
    func card(padding: CGFloat = 14, radius: CGFloat = CD.rCard) -> some View {
        modifier(CardModifier(padding: padding, radius: radius))
    }
}

// MARK: - 膠囊按鈕

struct PillButton: View {
    enum Style { case accent, lemon, ghost }
    let title: String
    var icon: String? = nil
    var style: Style = .accent
    var action: () -> Void

    @SwiftUI.State private var pressed = false

    var body: some View {
        Button {
            Haptics.shared.light()
            action()
        } label: {
            HStack(spacing: 8) {
                if let icon { Image(systemName: icon).font(.system(size: 15, weight: .bold)) }
                Text(title).font(.cdTitle(15.5))
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, 14)
            .background(background)
            .foregroundStyle(foreground)
            .clipShape(Capsule())
        }
        .buttonStyle(PressScaleStyle())
    }

    private var background: Color {
        switch style {
        case .accent: CD.accent
        case .lemon: CD.lemon
        case .ghost: CD.surface2
        }
    }
    private var foreground: Color {
        switch style {
        case .accent, .lemon: CD.plumDeep
        case .ghost: CD.text
        }
    }
}

struct PressScaleStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.96 : 1)
            .animation(CD.ease, value: configuration.isPressed)
    }
}

// MARK: - 標籤

struct TagView: View {
    let text: String
    var color: Color = CD.accent
    var filled: Bool = false

    var body: some View {
        Text(text)
            .font(.cdBody(11, weight: .bold))
            .padding(.horizontal, 10)
            .padding(.vertical, 4)
            .background(filled ? color : color.opacity(0.16))
            .foregroundStyle(filled ? CD.plumDeep : color)
            .clipShape(Capsule())
    }
}

// MARK: - 區塊標題

struct SectionHeader: View {
    let title: String
    var trailing: String? = nil
    var onTap: (() -> Void)? = nil

    var body: some View {
        HStack(alignment: .firstTextBaseline) {
            Text(title).font(.cdTitle(15)).foregroundStyle(CD.text)
            Spacer()
            if let trailing {
                Button { onTap?() } label: {
                    Text(trailing).font(.cdBody(12, weight: .bold)).foregroundStyle(CD.accent)
                }
                .buttonStyle(.plain)
            }
        }
    }
}

// MARK: - 列表項

struct ListRow<Icon: View, Trailing: View>: View {
    let iconBackground: Color
    let title: String
    let subtitle: String
    @ViewBuilder var icon: Icon
    @ViewBuilder var trailing: Trailing

    var body: some View {
        HStack(spacing: 11) {
            ZStack {
                RoundedRectangle(cornerRadius: CD.rIcon, style: .continuous)
                    .fill(iconBackground)
                    .frame(width: 36, height: 36)
                icon
            }
            VStack(alignment: .leading, spacing: 2) {
                Text(title).font(.cdTitle(13.5)).foregroundStyle(CD.text)
                    .lineLimit(1)
                Text(subtitle).font(.cdBody(11.5, weight: .medium)).foregroundStyle(CD.text2)
                    .lineLimit(2)
            }
            Spacer(minLength: 6)
            trailing
        }
        .padding(11)
        .background(CD.surface)
        .clipShape(RoundedRectangle(cornerRadius: CD.rRow, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: CD.rRow, style: .continuous)
                .stroke(CD.cardBorder, lineWidth: 1)
        )
    }
}

// MARK: - 進度條

struct SunProgressBar: View {
    let value: Double   // 0...1
    var body: some View {
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                Capsule().fill(CD.surface3)
                Capsule().fill(CD.lemon)
                    .frame(width: max(8, geo.size.width * value))
            }
        }
        .frame(height: 8)
        .animation(CD.easeSlow, value: value)
    }
}

// MARK: - 鎖定模糊遮罩（付費導流）

struct BlurLock<Content: View>: View {
    let cta: String
    @ViewBuilder var content: Content

    var body: some View {
        ZStack {
            content
                .blur(radius: 5)
                .opacity(0.7)
                .accessibilityHidden(true)
            Button {
                Haptics.shared.soft()   // 鎖定功能：soft 警示
            } label: {
                HStack(spacing: 6) {
                    Image(systemName: "lock.fill").font(.system(size: 11, weight: .bold))
                    Text(cta).font(.cdBody(12, weight: .heavy))
                }
                .padding(.horizontal, 16)
                .padding(.vertical, 9)
                .background(CD.accent)
                .foregroundStyle(CD.plumDeep)
                .clipShape(Capsule())
            }
            .buttonStyle(PressScaleStyle())
        }
    }
}

// MARK: - 頁面標題

struct PageHeader: View {
    let kicker: String
    let title: String

    var body: some View {
        VStack(alignment: .leading, spacing: 3) {
            Text(kicker).font(.cdBody(12, weight: .bold)).foregroundStyle(CD.text2)
            Text(title).font(.cdDisplay(24)).foregroundStyle(CD.text)
                .tracking(-0.5)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

// MARK: - 免責聲明

struct DisclaimerFooter: View {
    var body: some View {
        Text("此內容由 AI 輔助整理，請與原始醫療文件核對")
            .font(.cdBody(10.5, weight: .medium))
            .foregroundStyle(CD.text3)
            .frame(maxWidth: .infinity)
            .multilineTextAlignment(.center)
            .padding(.vertical, 8)
    }
}

File: flutter/carrius/lib/design/components.dart
import 'dart:ui';
import 'package:flutter/material.dart';
import 'tokens.dart';
import 'haptics.dart';

/// 卡片容器
class CDCard extends StatelessWidget {
  final Widget child;
  final EdgeInsets padding;
  final double radius;
  final Color? color;
  const CDCard({
    super.key,
    required this.child,
    this.padding = const EdgeInsets.all(14),
    this.radius = CD.rCard,
    this.color,
  });

  @override
  Widget build(BuildContext context) {
    final p = PaletteScope.of(context);
    return Container(
      padding: padding,
      decoration: BoxDecoration(
        color: color ?? p.surface,
        borderRadius: BorderRadius.circular(radius),
        border: Border.all(color: p.cardBorder, width: 1),
      ),
      child: child,
    );
  }
}

/// 膠囊按鈕
enum PillStyle { accent, lemon, ghost }

class PillButton extends StatefulWidget {
  final String title;
  final IconData? icon;
  final PillStyle style;
  final VoidCallback onTap;
  final bool dimmed;
  const PillButton({
    super.key,
    required this.title,
    this.icon,
    this.style = PillStyle.accent,
    required this.onTap,
    this.dimmed = false,
  });

  @override
  State<PillButton> createState() => _PillButtonState();
}

class _PillButtonState extends State<PillButton> {
  bool _pressed = false;

  @override
  Widget build(BuildContext context) {
    final p = PaletteScope.of(context);
    Color bg;
    Color fg;
    switch (widget.style) {
      case PillStyle.accent:
        bg = CD.accent;
        fg = CD.plumDeep;
      case PillStyle.lemon:
        bg = CD.lemon;
        fg = CD.plumDeep;
      case PillStyle.ghost:
        bg = p.surface2;
        fg = p.text;
    }
    return GestureDetector(
      onTapDown: (_) => setState(() => _pressed = true),
      onTapUp: (_) => setState(() => _pressed = false),
      onTapCancel: () => setState(() => _pressed = false),
      onTap: () {
        Haptics.light();
        widget.onTap();
      },
      child: AnimatedScale(
        scale: _pressed ? 0.96 : 1,
        duration: CD.easeDur,
        curve: CD.ease,
        child: Opacity(
          opacity: widget.dimmed ? 0.5 : 1,
          child: Container(
            width: double.infinity,
            padding: const EdgeInsets.symmetric(vertical: 14),
            decoration: BoxDecoration(color: bg, borderRadius: BorderRadius.circular(999)),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                if (widget.icon != null) ...[
                  Icon(widget.icon, size: 16, color: fg),
                  const SizedBox(width: 8),
                ],
                Text(widget.title, style: CDText.title(15.5, color: fg)),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

/// 標籤
class TagView extends StatelessWidget {
  final String text;
  final Color color;
  final bool filled;
  const TagView({super.key, required this.text, this.color = CD.accent, this.filled = false});

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 4),
      decoration: BoxDecoration(
        color: filled ? color : color.withValues(alpha: 0.16),
        borderRadius: BorderRadius.circular(999),
      ),
      child: Text(text,
          style: CDText.body(11, weight: FontWeight.w700, color: filled ? CD.plumDeep : color)),
    );
  }
}

/// 區塊標題
class SectionHeader extends StatelessWidget {
  final String title;
  final String? trailing;
  final VoidCallback? onTap;
  const SectionHeader({super.key, required this.title, this.trailing, this.onTap});

  @override
  Widget build(BuildContext context) {
    final p = PaletteScope.of(context);
    return Row(
      crossAxisAlignment: CrossAxisAlignment.baseline,
      textBaseline: TextBaseline.alphabetic,
      children: [
        Text(title, style: CDText.title(15, color: p.text)),
        const Spacer(),
        if (trailing != null)
          GestureDetector(
            onTap: () {
              Haptics.light();
              onTap?.call();
            },
            child: Text(trailing!, style: CDText.body(12, weight: FontWeight.w700, color: CD.accent)),
          ),
      ],
    );
  }
}

/// 列表項
class ListRowCard extends StatelessWidget {
  final Color iconBg;
  final Widget icon;
  final String title;
  final String subtitle;
  final Widget? trailing;
  final VoidCallback? onTap;
  const ListRowCard({
    super.key,
    required this.iconBg,
    required this.icon,
    required this.title,
    required this.subtitle,
    this.trailing,
    this.onTap,
  });

  @override
  Widget build(BuildContext context) {
    final p = PaletteScope.of(context);
    final row = Container(
      padding: const EdgeInsets.all(11),
      decoration: BoxDecoration(
        color: p.surface,
        borderRadius: BorderRadius.circular(CD.rRow),
        border: Border.all(color: p.cardBorder, width: 1),
      ),
      child: Row(
        children: [
          Container(
            width: 36,
            height: 36,
            decoration: BoxDecoration(color: iconBg, borderRadius: BorderRadius.circular(CD.rIcon)),
            child: Center(child: icon),
          ),
          const SizedBox(width: 11),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              mainAxisSize: MainAxisSize.min,
              children: [
                Text(title, style: CDText.title(13.5, color: p.text), maxLines: 1, overflow: TextOverflow.ellipsis),
                const SizedBox(height: 2),
                Text(subtitle,
                    style: CDText.body(11.5, weight: FontWeight.w500, color: p.text2),
                    maxLines: 2,
                    overflow: TextOverflow.ellipsis),
              ],
            ),
          ),
          if (trailing != null) ...[const SizedBox(width: 6), trailing!],
        ],
      ),
    );
    if (onTap == null) return row;
    return GestureDetector(
      onTap: () {
        Haptics.light();
        onTap!();
      },
      child: row,
    );
  }
}

/// 陽光進度條
class SunProgressBar extends StatelessWidget {
  final double value;
  const SunProgressBar({super.key, required this.value});

  @override
  Widget build(BuildContext context) {
    final p = PaletteScope.of(context);
    return ClipRRect(
      borderRadius: BorderRadius.circular(999),
      child: Container(
        height: 8,
        color: p.surface3,
        child: FractionallySizedBox(
          alignment: Alignment.centerLeft,
          widthFactor: value.clamp(0.02, 1),
          child: Container(color: CD.lemon),
        ),
      ),
    );
  }
}

/// 頁面標題
class PageHeader extends StatelessWidget {
  final String kicker;
  final String title;
  const PageHeader({super.key, required this.kicker, required this.title});

  @override
  Widget build(BuildContext context) {
    final p = PaletteScope.of(context);
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(kicker, style: CDText.body(12, weight: FontWeight.w700, color: p.text2)),
        const SizedBox(height: 3),
        Text(title, style: CDText.display(24, color: p.text)),
      ],
    );
  }
}

/// 鎖定模糊遮罩（付費導流）。點 CTA 走 onTap（通常接 showComingSoon），不可只震動不做事。
class BlurLock extends StatelessWidget {
  final String cta;
  final Widget child;
  final VoidCallback? onTap;
  const BlurLock({super.key, required this.cta, required this.child, this.onTap});

  @override
  Widget build(BuildContext context) {
    return Stack(
      alignment: Alignment.center,
      children: [
        ClipRRect(
          borderRadius: BorderRadius.circular(CD.rCard),
          child: ImageFiltered(
            imageFilter: ImageFilter.blur(sigmaX: 5, sigmaY: 5),
            child: Opacity(opacity: 0.7, child: child),
          ),
        ),
        GestureDetector(
          onTap: () {
            Haptics.light();
            onTap?.call();
          },
          child: Container(
            padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 9),
            decoration: BoxDecoration(color: CD.accent, borderRadius: BorderRadius.circular(999)),
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                const Icon(Icons.lock, size: 11, color: CD.plumDeep),
                const SizedBox(width: 6),
                Text(cta, style: CDText.body(12, weight: FontWeight.w900, color: CD.plumDeep)),
              ],
            ),
          ),
        ),
      ],
    );
  }
}

/// 免責聲明
class DisclaimerFooter extends StatelessWidget {
  const DisclaimerFooter({super.key});

  @override
  Widget build(BuildContext context) {
    final p = PaletteScope.of(context);
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 8),
      child: Text('此內容由 AI 輔助整理，請與原始醫療文件核對',
          textAlign: TextAlign.center, style: CDText.body(10.5, weight: FontWeight.w500, color: p.text3)),
    );
  }
}

/// 誠實「即將開放」提示（floating SnackBar，留出底部 TabBar 空間）。
/// 觸覺由呼叫端負責（PillButton/BlurLock/各 GestureDetector 自帶），這裡不重複觸發。
void showComingSoon(BuildContext context, String feature) {
  ScaffoldMessenger.of(context)
    ..clearSnackBars()
    ..showSnackBar(
      SnackBar(
        behavior: SnackBarBehavior.floating,
        backgroundColor: CD.plumDeep,
        elevation: 0,
        duration: const Duration(milliseconds: 2200),
        margin: const EdgeInsets.fromLTRB(18, 0, 18, 92),
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(14)),
        content: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            const Icon(Icons.schedule, size: 15, color: CD.cream),
            const SizedBox(width: 9),
            Flexible(
              child: Text('「$feature」即將在正式版開放',
                  style: CDText.body(12.5, weight: FontWeight.w700, color: CD.cream)),
            ),
          ],
        ),
      ),
    );
}

/// 底部資訊 sheet（隱私政策／同意書／方案／關於／文件詳情）
Future<void> showCDSheet(BuildContext context, {required String title, required Widget body}) {
  final p = PaletteScope.of(context);
  return showModalBottomSheet<void>(
    context: context,
    isScrollControlled: true,
    backgroundColor: Colors.transparent,
    builder: (ctx) {
      final media = MediaQuery.of(ctx);
      return Container(
        constraints: BoxConstraints(maxHeight: media.size.height * 0.82),
        decoration: BoxDecoration(
          color: p.surface,
          borderRadius: const BorderRadius.vertical(top: Radius.circular(22)),
        ),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            const SizedBox(height: 10),
            Container(
              width: 38,
              height: 4,
              decoration: BoxDecoration(color: p.surface3, borderRadius: BorderRadius.circular(999)),
            ),
            Padding(
              padding: const EdgeInsets.fromLTRB(20, 12, 12, 4),
              child: Row(
                children: [
                  Expanded(child: Text(title, style: CDText.title(16, color: p.text))),
                  GestureDetector(
                    key: const Key('cdSheetClose'),
                    onTap: () {
                      Haptics.light();
                      Navigator.of(ctx).pop();
                    },
                    child: Container(
                      width: 34,
                      height: 34,
                      decoration: BoxDecoration(color: p.surface2, shape: BoxShape.circle),
                      child: Icon(Icons.close, size: 15, color: p.text2),
                    ),
                  ),
                ],
              ),
            ),
            Flexible(
              child: SingleChildScrollView(
                padding: EdgeInsets.fromLTRB(20, 4, 20, 20 + media.padding.bottom),
                child: body,
              ),
            ),
          ],
        ),
      );
    },
  );
}

/// sheet 內文：小節標題
class SheetHeading extends StatelessWidget {
  final String text;
  const SheetHeading(this.text, {super.key});
  @override
  Widget build(BuildContext context) {
    final p = PaletteScope.of(context);
    return Padding(
      padding: const EdgeInsets.only(top: 14, bottom: 6),
      child: Text(text, style: CDText.title(13.5, color: p.text)),
    );
  }
}

/// sheet 內文：段落
class SheetParagraph extends StatelessWidget {
  final String text;
  const SheetParagraph(this.text, {super.key});
  @override
  Widget build(BuildContext context) {
    final p = PaletteScope.of(context);
    return Padding(
      padding: const EdgeInsets.only(bottom: 4),
      child: Text(text, style: CDText.body(13, weight: FontWeight.w500, color: p.text2, height: 1.5)),
    );
  }
}

/// sheet 內文：項目列（圓點）
class SheetBullet extends StatelessWidget {
  final String text;
  const SheetBullet(this.text, {super.key});
  @override
  Widget build(BuildContext context) {
    final p = PaletteScope.of(context);
    return Padding(
      padding: const EdgeInsets.only(top: 6),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Padding(
            padding: const EdgeInsets.only(top: 6, right: 9),
            child: Container(width: 5, height: 5, decoration: const BoxDecoration(color: CD.accent, shape: BoxShape.circle)),
          ),
          Expanded(child: Text(text, style: CDText.body(13, weight: FontWeight.w500, color: p.text2, height: 1.5))),
        ],
      ),
    );
  }
}

/// 流程頂欄（modal / sheet 用）
class FlowTopBar extends StatelessWidget {
  final String title;
  final VoidCallback onClose;
  const FlowTopBar({super.key, required this.title, required this.onClose});

  @override
  Widget build(BuildContext context) {
    final p = PaletteScope.of(context);
    return Padding(
      padding: const EdgeInsets.fromLTRB(18, 14, 18, 0),
      child: Row(
        children: [
          Text(title, style: CDText.title(16, color: p.text)),
          const Spacer(),
          GestureDetector(
            onTap: () {
              Haptics.light();
              onClose();
            },
            child: Container(
              width: 34,
              height: 34,
              decoration: BoxDecoration(color: p.surface2, shape: BoxShape.circle),
              child: Icon(Icons.close, size: 15, color: p.text2),
            ),
          ),
        ],
      ),
    );
  }
}
